webidls/Browser.webidl

```diff
@@ -0,0 +1,115 @@
+
+  // Used to expose special JS APIs.
+  // For example, can be used to expose the Browser API or Web Extensions APIs for content
+  // See: https://developer.chrome.com/extensions/content_scripts#execution-environment
+  Promise<void> addJSModuleResolver(/*FIXME*/, boolean onlyForFrameScript);
+
+  void destroy();
+
+  // Can't change after browser creation
+  readonly attribute DOMString browsingContextName;
+  readonly attribute StorageSession storageSession;
+
+  // Will fail if entries already exist.
+  // This can be used for session restore, or to undo tab-close.
+  // On success, historyEntries is filled, active entry's pipeline is NOT pending.
+  // If necessary, other pipelines can be restored via HistoryEntry.restorePipeline()
+  Promise<void> restoreEntries(FrozenArray<LoadData> data, unsigned long activeIndex);
+
+  readonly attribute Sequence<HistoryEntry> historyEntries;
+  readonly attribute unsigned long activeEntryIndex;
+  Promise<void> setActiveEntryIndex(unsigned long);
+
+  // Use to load a new URL. will create a new pipeline and navigate to the
+  // pipeline once not pending
+  Promise<HistoryEntry> navigate(LoadData loadData, optional Pipeline opener);
+
+  // Will fail if Browser.session != pipeline.session.
+  // Useful with prerendering pipelines
+  Promise<Pipeline> createPrerenderingPipeline(LoadData loadData);
+  Promise<HistoryEntry> navigateToPipeline(Pipeline pipeline, optional Pipeline opener);
+
+
+  // FIXME: do we want to give access to autopurge and maxLivePipeline?
+  // Purging means killing the pipeline. The entry stays intact
+  readonly attribute boolean doPipelinesAutopurge;
+  // Maximum number of live pipeline before and after the current entry
+  Promise<void> enableAutoPurgePipelines(unsigned long before, unsigned long after);
+  Promise<void> disableAutoPurgePipelines();
+
+  // Popup blocker, tracking content blocker, mixed content blocker,
+  // and safari-like content blocker. See ContentBlockers.webidl
+  // FIXME: not cool. We can't have multiple custom blockers.
+  readonly attribute Sequence<ContentBlockerDescription> contentBlockers;
+  Promise<void> setContentBlockers(Sequence<ContentBlockerDescription>);
+
+  // FIXME: not sure focus should be handled here
+  readonly attribute boolean isFocused;
+
+  // FIXME: can we do without prefs?
+  readonly attribute Object prefs;
+  Promise<void> setPrefs(Object prefs); // use to set user-agent for example
+
+  Promise<boolean/*prevent default canceled*/> dispatchKeyEvent(KeyEvent event);
+}
+
+interface BrowserHandler {
+
+  // A new document is displayed. Usually after the user
+  // clicked on a link and once the new document has been
+  // created (pipeline is not pending anymore). Also happens
+  // when user or page goes back/forward.
+  void onActiveEntryChanged();
+
+  void onFocusChanged();
+  
+  // window.onClose() is called
+  Cancelable onDestroyRequestedFromContent();
+
+  void onDestroy();
```

Comment:
My guess is that this is called after destroy rather than before if so it should be `onDestroyed`.